// node-thermal-printer/lib/types/epson.js
const PrinterType = require('./printer-type');

class Epson extends PrinterType {
  constructor() {
    super();
    this.config = require('./epson-config');
  }

  // ------------------------------ Get paper status ------------------------------
  getStatus() {
    return this.config.TRANSMIT_PAPER_STATUS;
  }

  // ------------------------------ Append ------------------------------
  append(appendBuffer) {
    if (this.buffer) {
      this.buffer = Buffer.concat([this.buffer, appendBuffer]);
    } else {
      this.buffer = appendBuffer;
    }
  }

  // ------------------------------ Beep ------------------------------
  beep(numberOfBeeps = 1, lengthOfTheSound = 1) {
    if (numberOfBeeps < 1 || numberOfBeeps > 9) throw new Error('numberOfBeeps: Value must be between 1 and 9');
    if (lengthOfTheSound < 1 || lengthOfTheSound > 9) throw new Error('lengthOfTheSound: Value must be between 1 and 9');
    this.buffer = null;
    this.append(this.config.BEEP);
    this.append(Buffer.from([numberOfBeeps, lengthOfTheSound]));
    return this.buffer;
  }

  // ------------------------------ Set text size ------------------------------
  setTextSize(height, width) {
    this.buffer = null;
    if (height > 7 || height < 0) throw new Error('setTextSize: Height must be between 0 and 7');
    if (width > 7 || width < 0) throw new Error('setTextSize: Width must be between 0 and 7');
    const x = Buffer.from(`${height}${width}`, 'hex');
    this.append(Buffer.from([0x1D, 0x21]));
    this.append(x);
    return this.buffer;
  }

  // ------------------------------ CODE 128 ------------------------------
  code128(data, settings) {
    this.buffer = null;
    settings = {
      hriPos: 0,
      hriFont: 0,
      width: 3,
      height: 162,
      ...settings
    };

    this.append(Buffer.from([0x1d, 0x48])); // GS H
    this.append(Buffer.from([settings.hriPos]));
    this.append(Buffer.from([0x1d, 0x66])); // GS f
    this.append(Buffer.from([settings.hriFont]));
    this.append(Buffer.from([0x1d, 0x77])); // GS W
    this.append(Buffer.from([settings.width]));
    this.append(Buffer.from([0x1d, 0x68])); // GS h
    this.append(Buffer.from([settings.height]));
    this.append(this.config.BARCODE_CODE128);
    this.append(Buffer.from([data.length + 2]));
    this.append(Buffer.from([0x7b, 0x42]));
    this.append(Buffer.from(data));
    return this.buffer;
  }

  // ------------------------------ QR ------------------------------
  printQR(str, settings) {
    this.buffer = null;
    settings = {
      model: 2,
      cellSize: 3,
      correction: 'M',
      ...settings
    };

    if (settings.model === 1) this.append(this.config.QRCODE_MODEL1);
    else if (settings.model === 3) this.append(this.config.QRCODE_MODEL3);
    else this.append(this.config.QRCODE_MODEL2);

    const size = 'QRCODE_CELLSIZE_'.concat(settings.cellSize.toString());
    this.append(this.config[size]);

    const correction = 'QRCODE_CORRECTION_'.concat(settings.correction.toUpperCase());
    this.append(this.config[correction]);

    const s = str.length + 3;
    const lsb = parseInt(s % 256);
    const msb = parseInt(s / 256);
    this.append(Buffer.from([0x1d, 0x28, 0x6b, lsb, msb, 0x31, 0x50, 0x30]));
    this.append(Buffer.from(str));
    this.append(this.config.QRCODE_PRINT);

    return this.buffer;
  }

  // ------------------------------ PDF417 ------------------------------
  pdf417(data, settings) {
    this.buffer = null;
    settings = {
      correction: 1,
      rowHeight: 3,
      width: 3,
      columns: 0,
      truncated: false,
      ...settings
    };

    this.append(this.config.PDF417_CORRECTION);
    this.append(Buffer.from([settings.correction]));
    this.append(this.config.PDF417_ROW_HEIGHT);
    this.append(Buffer.from([settings.rowHeight]));
    this.append(this.config.PDF417_WIDTH);
    this.append(Buffer.from([settings.width]));
    this.append(this.config.PDF417_COLUMNS);
    this.append(Buffer.from([settings.columns]));
    if (settings.truncated) this.append(this.config.PDF417_OPTION_TRUNCATED);
    else this.append(this.config.PDF417_OPTION_STANDARD);

    const s = data.length + 3;
    const lsb = parseInt(s % 256);
    const msb = parseInt(s / 256);
    this.append(Buffer.from([0x1d, 0x28, 0x6b, lsb, msb, 0x30, 0x50, 0x30]));
    this.append(Buffer.from(data.toString()));
    this.append(Buffer.from(this.config.PDF417_PRINT));

    return this.buffer;
  }

  // ------------------------------ MAXI CODE ------------------------------
  maxiCode(data, settings) {
    this.buffer = null;
    settings = {
      mode: 4,
      ...settings
    };

    if (settings.mode == 2) this.append(this.config.MAXI_MODE2);
    else if (settings.mode == 3) this.append(this.config.MAXI_MODE3);
    else if (settings.mode == 5) this.append(this.config.MAXI_MODE5);
    else if (settings.mode == 6) this.append(this.config.MAXI_MODE6);
    else this.append(this.config.MAXI_MODE4);

    const s = data.length + 3;
    const lsb = parseInt(s % 256);
    const msb = parseInt(s / 256);
    this.append(Buffer.from([0x1d, 0x28, 0x6b, lsb, msb, 0x32, 0x50, 0x30]));
    this.append(Buffer.from(data.toString()));
    this.append(this.config.MAXI_PRINT);

    return this.buffer;
  }

  // ------------------------------ BARCODE ------------------------------
  printBarcode(data, type, settings) {
    this.buffer = null;
    settings = {
      hriPos: 0,
      hriFont: 0,
      width: 3,
      height: 162,
      ...settings
    };

    this.append(Buffer.from([0x1d, 0x48])); // GS H
    this.append(Buffer.from([settings.hriPos]));
    this.append(Buffer.from([0x1d, 0x66])); // GS f
    this.append(Buffer.from([settings.hriFont]));
    this.append(Buffer.from([0x1d, 0x77])); // GS W
    this.append(Buffer.from([settings.width]));
    this.append(Buffer.from([0x1d, 0x68])); // GS h
    this.append(Buffer.from([settings.height]));
    this.append(Buffer.from([0x1d, 0x6b])); // GS k

    if (type == 73) {
      this.append(Buffer.from([type, data.length + 2]));
      this.append(Buffer.from([0x7b, 0x42]));
    } else {
      this.append(Buffer.from([type, data.length]));
    }
    this.append(Buffer.from(data));

    return this.buffer;
  }

  // ------------------------------ PRINT IMAGE ------------------------------
  async printImage(image) {
    const fs = require('fs');
    const { PNG } = require('pngjs');
    try {
      const data = fs.readFileSync(image);
      const png = PNG.sync.read(data);
      const buff = this.printImageBuffer(png.width, png.height, png.data);
      return buff;
    } catch (error) {
      throw error;
    }
  }

  printImageBuffer(width, height, data) {
    this.buffer = null;

    const pixels = [];
    for (let i = 0; i < height; i++) {
      const line = [];
      for (let j = 0; j < width; j++) {
        const idx = (width * i + j) << 2;
        line.push({
          r: data[idx],
          g: data[idx + 1],
          b: data[idx + 2],
          a: data[idx + 3],
        });
      }
      pixels.push(line);
    }

    const imageBufferArray = [];
    for (let i = 0; i < height; i++) {
      for (let j = 0; j < Math.ceil(width / 8); j++) {
        let byte = 0x0;
        for (let k = 0; k < 8; k++) {
          let pixel = pixels[i][j * 8 + k];
          if (pixel === undefined) {
            pixel = { a: 0, r: 0, g: 0, b: 0 };
          }
          if (pixel.a > 126) {
            const grayscale = parseInt(0.2126 * pixel.r + 0.7152 * pixel.g + 0.0722 * pixel.b);
            if (grayscale < 128) {
              const mask = 1 << (7 - k);
              byte |= mask;
            }
          }
        }
        imageBufferArray.push(byte);
      }
    }

    const imageBuffer = Buffer.from(imageBufferArray);

    if (width % 8 != 0) {
      width += 8;
    }

    this.append(Buffer.from([0x1d, 0x76, 0x30, 48]));
    this.append(Buffer.from([(width >> 3) & 0xff]));
    this.append(Buffer.from([0x00]));
    this.append(Buffer.from([height & 0xff]));
    this.append(Buffer.from([(height >> 8) & 0xff]));
    this.append(imageBuffer);

    return this.buffer;
  }

  // ------------------------------ Set Character Set ------------------------------
  setCharacterSet(characterSet) {
    console.log(`Setting character set: ${characterSet}`);
    const buffer = this.config[`CODE_PAGE_${characterSet}`];
    if (buffer) {
      this.append(buffer);
      console.log(`Character set ${characterSet} applied successfully: ${buffer.toString('hex')}`);
      return true;
    } else {
      console.error(`Character set ${characterSet} not found in config`);
      // Fallback to raw ESC/POS command for Chinese
      if (characterSet === 'CHINA') {
        console.log('Falling back to raw ESC t 48 for CHINA (PC936, GBK)');
        this.append(Buffer.from([0x1B, 0x74, 0x30])); // ESC t 48 (PC936, GBK)
      } else if (characterSet === 'HK_TW') {
        console.log('Falling back to raw ESC t 50 for HK_TW (PC950, Big5)');
        this.append(Buffer.from([0x1B, 0x74, 0x32])); // ESC t 50 (PC950, Big5)
      } else {
        throw new Error(`Code page not recognized: '${characterSet}'`);
      }
    }
  }
}

module.exports = Epson;
